lib/AST/DiagnosticEngine.cpp

```diff
@@ -385,6 +385,25 @@ static bool shouldShowAKA(Type type, StringRef typeName) {
   return true;
 }
 
+/// If a type is part of an argument list which includes another, distinct type
+/// with the same string representation, it should be qualified during
+/// formatting.
+static bool shouldQualifyType(Type type, ArrayRef<DiagnosticArgument> Args) {
+  if (!type->isCanonicalizable()) {
+    return false;
+  }
+  for (auto arg : Args) {
+    if (arg.getKind() == DiagnosticArgumentKind::Type) {
+      auto argType = arg.getAsType();
+      if (argType && argType->isCanonicalizable() && !argType->isEqual(type) &&
+          argType->getWithoutParens().getString() == type.getString()) {
```

Comment:
What's an example where the two types only differ by parens but are otherwise equal?

Reply:
A little bit of a contrived example, but the following only qualifies both types when diagnosed if the parens are stripped:

```swift
struct Error {}
let baz: Swift.Error = (Error())
```

I added a test case for it.

Reply:
Shouldn't you strip parens from both sides then? What if you declare 'let baz: ((Swift.Error))' or something here

Reply:
`formatDiagnosticArgument` strips the parens from `type` in advance before printing, so it isn't necessary to do it a second time in `shouldQualifyType`. The example you give is diagnosed with both types qualified as expected.

Reply:
Ok, great. Can you add a test case (or make sure one already exists)

Reply:
done, added a couple cases with different levels of nested parens